### PR TITLE
Kickoff: Add watchdog for loop interactivity.

### PR DESCRIPTION
**Pull request by PremyslCerny**
Discussed in Slack.

### CODE REVIEW

**Review by mejroslav — APPROVED** [review]
Let's try it!

File: asab/application.py [hunk 1]
```diff
@@ -130,6 +130,12 @@ def __init__(self):
 		self.LaunchTime = time.time()
 		self.BaseTime = self.LaunchTime - self.Loop.time()
 
+		# Last time the on_tick/60! was registered
+		self.LastOnTick60 = self.LaunchTime
+
+		# Variable that is set when the exit time is happening
+		self.Exiting = False
```
Comment (ateska) [inline asab/application.py:137]: Definitively not - no state variables on this level.

File: asab/application.py [hunk 2]
```diff
@@ -185,6 +191,12 @@ def handler(type):
 
 		self.TaskService = TaskService(self)
 
+		# Proactor is mandatory for subsequent interactivity check
+		import asab.proactor
```
Comment (ateska) [inline asab/application.py:195]: Nope, use `threading` from Python.

**Review by ateska — CHANGES_REQUESTED** [review]
Be reasonable with the approach, this is a core of all microservices.

**Comment by PremyslCerny** [comment]
I tested it with Correlator so far. It does not influence the init, run or exit time performance. Still, since it will influence or msvs, I would like to be more sure.

File: asab/application.py [hunk 3]
```diff
@@ -194,6 +199,28 @@ def handler(type):
 		# Every 10 minutes listen for housekeeping
 		self.PubSub.subscribe("Application.tick/600!", self._on_housekeeping_tick)
 
+		# Run the watchdog to detect lost interactivity on the event loop
+		thread = threading.Thread(target=self._run_watchdog_for_event_loop)
```
Comment (ateska) [inline asab/application.py:203]: `watchdog_thread = `

File: asab/application.py [hunk 4]
```diff
@@ -194,6 +199,28 @@ def handler(type):
...
+		thread.daemon = True  # Daemonize the thread to ensure it exits with the main program
+		thread.start()
+
+
+	def _run_watchdog_for_event_loop(self):
```
Comment (ateska) [inline asab/application.py:208]: `def _watchdog(self):`

... don't try to squeeze documentation into the name of the method, please.

File: asab/application.py [hunk 5]
```diff
@@ -194,6 +199,28 @@ def handler(type):
...
+		while True:
+			current_time = time.time()
+
+			if (current_time - self.LastOnTick60) < (15 * 60):  # Fifteen minute threshold (15 * expected cycle)
+				time.sleep(60)  # Sleep one minute (one cycle)
+				continue
+
+			# The loop lost its interactivity
+			L.critical("The event loop lost its interactivity. Stopping the application!")
+			os._exit(5)  # Cannot use sys.exit inside the thread
```
Comment (ateska) [inline asab/application.py:222]: Why?

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/application.py (hunk 5 above)
Comment (PremyslCerny) [reply asab/application.py:222]: Because sys.exit simply raises SystemExit, which basically terminates the thread but not the application.

Comment (ateska) [reply asab/application.py:222]: consider killing the process then using: `os.kill(os.getpid(), signal.SIGKILL)`

File: asab/application.py [hunk 6]
```diff
@@ -194,6 +199,28 @@ def handler(type):
...
+
+	def _watchdog(self):
+		"""
+		Periodically checks the loop interactivity.
+		"""
+
+		while True:
+			current_time = time.time()
+
+			if (current_time - self.LastOnTick60) < (15 * 60):  # Fifteen minute threshold (15 * expected cycle)
```
Comment (ateska) [inline asab/application.py:216]: `15 * 60` must be configurable, with default and possibility to disable this check completely.

**Review by ateska — CHANGES_REQUESTED** [review]

File: asab/application.py [hunk 7]
```diff
@@ -194,6 +199,28 @@ def handler(type):
...
 
+		# Run the watchdog to detect lost interactivity on the event loop
+		watchdog_thread = threading.Thread(target=self._watchdog)
+		watchdog_thread.daemon = True  # Daemonize the thread to ensure it exits with the main program
+		watchdog_thread.start()
+
+
+	def _watchdog(self):
+		"""
+		Periodically checks the loop interactivity.
```
Comment (ateska) [inline asab/application.py:210]: ... and maily KILLing a stuck process by a signal - that's an important bit to say

**Review by ateska — CHANGES_REQUESTED** [review]

**Review by ateska — APPROVED** [review]

**Review by mejroslav — APPROVED** [review]

**Comment by PremyslCerny** [comment]
@mejroslav Can you please test it with parsec if you have time? You should not observe any change.

**Pull request merged**